samples/Common.c

```diff
@@ -999,3 +915,132 @@ STATUS sessionCleanupWait(PSampleConfiguration pSampleConfiguration)
     LEAVES();
     return retStatus;
 }
+
+STATUS submitPendingIceCandidate(PStackQueue pPendingMessageQueue, PSampleStreamingSession pSampleStreamingSession)
+{
+    STATUS retStatus = STATUS_SUCCESS;
+    BOOL noPendingSignalingMessageForClient = FALSE;
+    PReceivedSignalingMessage pReceivedSignalingMessage = NULL;
+
+    do {
+        CHK_STATUS(stackQueueIsEmpty(pPendingMessageQueue, &noPendingSignalingMessageForClient));
+        if (!noPendingSignalingMessageForClient) {
+            CHK_STATUS(stackQueueDequeue(pPendingMessageQueue, (PUINT64) &pReceivedSignalingMessage));
+            if (pReceivedSignalingMessage->signalingMessage.messageType == SIGNALING_MESSAGE_TYPE_ICE_CANDIDATE) {
+                CHK_STATUS(handleRemoteCandidate(pSampleStreamingSession, &pReceivedSignalingMessage->signalingMessage));
+            }
+            MEMFREE(pReceivedSignalingMessage);
+        }
+    } while (!noPendingSignalingMessageForClient);
+    CHK_STATUS(stackQueueFree(pPendingMessageQueue));
+
+CleanUp:
+
+    CHK_LOG_ERR(retStatus);
+    return retStatus;
+}
+
+STATUS signalingMessageReceived(UINT64 customData, PReceivedSignalingMessage pReceivedSignalingMessage)
+{
+    STATUS retStatus = STATUS_SUCCESS;
+    PSampleConfiguration pSampleConfiguration = (PSampleConfiguration) customData;
+    BOOL peerConnectionFound = FALSE;
+    BOOL locked = TRUE;
+    UINT32 clientIdHash;
+    PStackQueue pPendingMessageQueue = NULL;
+    PSampleStreamingSession pSampleStreamingSession = NULL;
+    PReceivedSignalingMessage pReceivedSignalingMessageCopy = NULL;
+
+    CHK(pSampleConfiguration != NULL, STATUS_NULL_ARG);
+
+    MUTEX_LOCK(pSampleConfiguration->sampleConfigurationObjLock);
+    locked = TRUE;
+
+    clientIdHash = COMPUTE_CRC32((PBYTE) pReceivedSignalingMessage->signalingMessage.peerClientId,
+                                 (UINT32) STRLEN(pReceivedSignalingMessage->signalingMessage.peerClientId));
+    CHK_STATUS(hashTableContains(pSampleConfiguration->pRtcPeerConnectionForRemoteClient, clientIdHash, &peerConnectionFound));
+    if (peerConnectionFound) {
+        CHK_STATUS(hashTableGet(pSampleConfiguration->pRtcPeerConnectionForRemoteClient, clientIdHash, (PUINT64) &pSampleStreamingSession));
+    }
+
+    switch (pReceivedSignalingMessage->signalingMessage.messageType) {
+        case SIGNALING_MESSAGE_TYPE_OFFER:
+            /*
+             * Create new streaming session for each offer, then insert the client id and streaming session into
+             * pRtcPeerConnectionForRemoteClient for subsequent ice candidate messages. Lastly check if there is
+             * any ice candidate messages queued in pPendingSignalingMessageForRemoteClient. If so then submit
+             * all of them.
+             */
+            CHK_STATUS(createSampleStreamingSession(pSampleConfiguration, pReceivedSignalingMessage->signalingMessage.peerClientId, TRUE,
```

Comment:
Since we are refactoring the sample, we should think on how to make this operation prompt. The main issue here is that it can block on the awaiting for ice config which then blocks the callback thread that executes this API. However, this is OK as it's a separate thread just for the message delivery and it in fact assumes non-prompt operations could happen so we should be OK.